### PR TITLE
Breaking: rule no-empty also checking for empty catch blocks. (fixes #1841)

### LINKED ISSUE #1841: Option to have no-empty rule flag empty 'catch' blocks
Many times when just prototyping for the first time, folks will use this construct because they're not interested in the error:

```js
try {
... code that code could throw
} catch (e) {
}
```

But later, for production-ready code, you want to make sure that there is indeed error handling code in those 'catch' blocks.

The current rule doesn't flag 'catch' blocks. I'd like to propose a config option for the 'no-empty' rule that would flag them.

Maybe something like: `"no-empty": [2, {"includeCatchBlocks": true}]`or some such.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/8624737-option-to-have-no-empty-rule-flag-empty-catch-blocks?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by doberkofler**

### CODE REVIEW

File: lib/rules/no-empty.js [hunk 1]
```diff
@@ -14,20 +15,39 @@ module.exports = function(context) {
...
+                i;
+
+            // if the body is not empty, we can just return immediately
+            if (node.body.length !== 0) {
                 return;
             }
 
-            if (node.body.length === 0) {
-                context.report(node, "Empty block statement.");
+            ancestors = context.getAncestors();
```
Comment (nzakas) [inline lib/rules/no-empty.js]: You can just do `node.parent` to get the parent.

File: lib/rules/no-empty.js [hunk 2]
```diff
@@ -14,20 +15,39 @@ module.exports = function(context) {
...
+            parentType = parent.type;
+            isFinallyBlock = (parentType === "TryStatement") && (parent.finalizer === node);
+
+            // a function is allowewd to be empty
+            if (parentType === "FunctionDeclaration" || parentType === "FunctionExpression" || parentType === "ArrowFunctionExpression") {
+                return;
             }
+
+            // catch and finally in a try statement are only allowed to be empty, if they contain an empty comment
+            if (parentType === "CatchClause" || isFinallyBlock) {
```
Comment (nzakas) [inline lib/rules/no-empty.js]: This exception isn't just for `catch` and `finally`, it's for any empty block.

File: lib/rules/no-empty.js (hunk 1 above)
Comment (doberkofler) [inline lib/rules/no-empty.js]: I actually kept this from your original code but clearly only using node.parent seems more efficient.

**Comment by nzakas** [comment]
This looks great, thanks!

**Pull request merged**